finality-aleph/src/sync/handler/mod.rs

```diff
@@ -643,6 +643,43 @@ where
         (new_highest, None)
     }
 
+    /// Processes the given `header` passed by `peer`. May return an error which indicates that the header couldn't
+    /// have been verified correctly, forest couldn't have been updated with the header, or the corresponding block
+    /// is not needed to be imported. If the block is not importable from the perspective of the forest, we can
+    /// ignore it if we detect that the handler processes a sequence of block imports which is indicated by
+    /// `last_imported` being the id of the parent of the block of the `header`.
+    fn handle_header(
+        &mut self,
+        header: UnverifiedHeaderFor<J>,
+        peer: I,
+        equivocation_proofs: &mut Vec<V::EquivocationProof>,
+        last_imported: Option<BlockId>,
+    ) -> Result<(), <Self as HandlerTypes>::Error> {
+        let h = match self.verify_header(header, false) {
+            Ok(VerifiedHeader {
+                header: h,
+                maybe_equivocation_proof,
+            }) => {
+                if let Some(proof) = maybe_equivocation_proof {
+                    equivocation_proofs.push(proof);
+                }
+                h
+            }
+            Err(e) => return Err(e),
+        };
+        if let Err(e) = self.forest.update_header(&h, Some(peer), false) {
+            return Err(Error::Forest(e));
+        }
+        if !self.forest.importable(&h.id())
+            && !last_imported
+                .map(|id| h.parent_id().map(|p_id| id == p_id).unwrap_or(false))
```

Comment:
`flatten` instead of the double `unwrap_or`? Or rather `flat_map`.

Reply:
Sorry, I don't see how to use `flat_map` to simplify this... :thinking:

Reply:
Change the `map` at the beginning of this line into `flat_map` and remove the `unwrap_or` inside the closure.

Reply:
`Option` doesn't support `flat_map`.

Reply:
I think `and_then` has similar semantics, let me try this